src/react-components/input/Slider.js

```diff
@@ -0,0 +1,143 @@
+export function Slider({ min, max, step, digits, value, defaultValue, onChange, disabled, className }) {
+  const myRoot = useRef();
+  const inputRef = useRef();
+  const [displayValue, setDisplayValue] = useState(-1);
+  const [isDragging, setIsDragging] = useState(false);
+
+  let currentValue = defaultValue;
+  if (isDragging) {
+    currentValue = displayValue !== -1 ? displayValue : defaultValue;
+  } else {
+    currentValue = value !== undefined ? value : displayValue !== -1 ? displayValue : defaultValue;
+  }
+
+  const stopDragging = useCallback(
+    () => {
+      setIsDragging(false);
+    },
+    [setIsDragging]
+  );
+
+  const onUpdateValue = useCallback(
+    v => {
+      setDisplayValue(v);
+      if (onChange) {
+        onChange(v);
+      }
+    },
+    [onChange, setDisplayValue]
+  );
+
+  const drag = useCallback(
+    e => {
+      if (!isDragging) return;
+      const viewportOffset = myRoot.current.getBoundingClientRect();
+      const t = Math.max(0, Math.min((e.clientX - viewportOffset.left) / viewportOffset.width, 1));
+      const num = round(step, min + t * (max - min));
+      onUpdateValue(num);
+    },
+    [isDragging, step, min, max, onUpdateValue]
+  );
+
+  useEffect(
+    () => {
+      window.addEventListener("mouseup", stopDragging);
+      window.addEventListener("mousemove", drag);
+
+      updateGradient(inputRef.current, min, max, currentValue);
+
+      return () => {
+        window.removeEventListener("mouseup", stopDragging);
+        window.removeEventListener("mousemove", drag);
+      };
+    },
+    [drag, stopDragging, inputRef, min, max, currentValue]
+  );
+
+  return (
+    <div className={classNames(styles.numberWithRange, className)}>
+      <div
+        ref={myRoot}
+        className={classNames(styles.rangeSlider)}
+        disabled={disabled}
+        onMouseDown={e => {
+          e.preventDefault();
+          if (disabled) return;
+          setIsDragging(true);
+          const viewportOffset = myRoot.current.getBoundingClientRect();
+          const t = Math.max(0, Math.min((e.clientX - viewportOffset.left) / viewportOffset.width, 1));
+          const num = round(step, min + t * (max - min));
+          const finalNum = parseFloat(num.toFixed(digits));
+          onUpdateValue(finalNum);
+        }}
+      >
+        <input
+          ref={inputRef}
+          type="range"
+          step={step}
+          min={min}
+          max={max}
+          value={currentValue}
+          disabled={disabled}
+          onChange={e => {
+            if (disabled) return;
+            const num = round(step, parseFloat(e.target.value));
+            const finalNum = parseFloat(num.toFixed(digits));
+            onUpdateValue(finalNum);
+          }}
+        />
+      </div>
+    </div>
+  );
+}
+
+Slider.propTypes = {
+  min: PropTypes.number,
+  max: PropTypes.number,
+  step: PropTypes.number,
+  digits: PropTypes.number,
+  value: PropTypes.number,
+  defaultValue: PropTypes.number,
```

Comment:
Having a `defaultValue` on a widget seems surprising. Its the responsibility of the thing using this component to provide a value to display, this component just tells you when the user tries to drag to change it, and then you pass back in the updated value.

Reply:
This still seems true (though is really just a carry through from this component having state being itself surprising). The main case I could see for a component like this having a default value would be if it internally handled "resetting" to some default, but even then I would probably just have it take an `onReset` function and have the thing using the component handle actually changing the value.